### ISSUE TITLE
Product Collection: Interactivity API - Handling "Interactive" blocks in client side navigation

### ISSUE THREAD

**Issue by kmanijak** [woothemes/woocommerce#43395]
**Background:**
Blocks that are interactive but not with Interactivity API (e.g. based on DOM API or jQuery) are not guaranteed to work correctly within Product Collection block. When the client-side navigation happens, there's no guarantee all the necessary events are fired etc.

**Goal:**
Possible ways to go:
1. Prevent the insertion of any non-compatible block in the Products Collection block. 
2. Allow the insertion of any block in the Products Collection block. Then, check the blocks in the server and if you find a non-compatible block, disable the client-side navigation.

I'm not sure if we can determine the 1. so I'd vote for option 2.

**Comment by sunyatasattva** [comment]
Hey @kmanijak! This was a problem we had known we were going to face. Here you can find a small summary of @luisherranz's breakdown: peCmuQ-1W-p2. Also see [this conversation](https://github.com/WordPress/gutenberg/discussions/54194) for the proposal of the compatibility declaration.

In a nutshell, Luis had proposed your solution 1., but in reverse: basically white-listing only compatible blocks, rather than blacklisting incompatible ones for the time being (until compatibility can be explicitly declared by blocks). Your proposed solution 2. is also worth discussing, depending on one factor: which blocks that we own would we currently disable if we were to take solution 1? How crucial are they?

**Comment by kmanijak** [comment]
Thanks @sunyatasattva!
> In a nutshell, Luis had proposed your solution 1., but in reverse: basically white-listing only compatible blocks, rather than blacklisting incompatible ones for the time being (until compatibility can be explicitly declared by blocks).

This is indeed what @luisherranz proposed but solution 2 is what is implemented in Query Loop block:

> Now, for example, when someone tries to activate https://github.com/WordPress/gutenberg/pull/53812, if all blocks that 
are inside the Post Template are compatible (they have supports.interactivity: true/false), the feature is activated without issues. If there's an unknown block, the user is notified that some specific block/s are incompatible, and the feature cannot be activated. If the feature is activated and the user introduces an unknown block in the Post Template, a message pops up saying that this block is not compatible and the feature has been deactivated.

Regarding your question:
> which blocks that we own would we currently disable if we were to take solution 1? How crucial are they?

I browsed the blocks that we own for the usage of DOM API or jQuery; they all seem safe (except for Mini Cart or Cart/Checkout but they're not intended to be used in Product Collection). My understanding is that we don't have control over other blocks (3PD blocks) that may be incompatible with Interactivity API.

What we could do is mark all the non-interactive blocks as Interactivity-compatible (as per discussion https://github.com/WordPress/gutenberg/discussions/54194) and whitelist only blocks explicitly declaring compatibility as per Luiz:
> If someone wants to create a block that always uses one of these features they can prevent users from adding not-compatible inner blocks in the Editor. That way, they will ensure that the features will always work and don't have to circumvent activating/deactivating the feature when a not-compatible block is added.

One downside of solution 1 is that limits the potential extensibility of Product Collection block.
The downside of solution 2 is the necessity to expose another option in Inspector Controls to toggle enhanced navigation which in my opinion should be just enabled.

**Comment by luisherranz** [comment]
I suggested option 1, which differs from what we did for the Query block, because as far as I know you didn't want to make the client-side navigation optional.

You can go with option 2 as well, but you need to make sure that all the filters and blocks support server-side navigation as well, and always test them in both scenarios.

Overall, the Interactivity API supports both paths, so you should choose the one you think is best for your use case without relying on what we did for the Query block.

By the way, one additional comment. At the time I did the initial PR (the one [Karol mentioned](https://github.com/WordPress/gutenberg/pull/53812)) we were trying to know if the blocks were compatible or not in the Editor. This was limiting because we had to rule out blocks that can be edited outside of the page that contains the Query block, like Patterns or Template parts. Imagine this:

- I add a Query block, enable client-side navigation.
- I add a Pattern inside, it doesn't contain non-supported blocks.
- Everything is fine, I save the page and the client-side navigation is enabled.
- I edit the Pattern on another page or in isolation.
- I add a non-supported block.
- Now the Query block contains non-supported blocks, but it has client-side navigation enabled.

At first, we started considering these types of blocks as non-supported, but people complained because TwentyTwentyFour was full of them, and that means people could not enable the client-side navigation on it. So, in the last minute, we allowed them, which means that turning client-side navigation ON is not a guarantee that it will work, and we now check in the server and turn it OFF automatically if we detect non-supported blocks at render time ([related PR](https://github.com/WordPress/gutenberg/pull/55714)).

In a way, this aligns better with our initial ideas about how client-side navigation should work in WordPress, with three settings:

- `off`: it's always turned off
- `auto`: it's turned on if everything is compatible (that's how it currently works in the Query block when you enable it)
- `on`: it's always turned on, use it at your own risk

What I mean with all this is that I think it's not possible or feasible to know in the Editor if the blocks are compatible or not if you want to include blocks like Patterns and Template Parts. So if you want to go with option 1, you need to take that into account.

**Comment by sunyatasattva** [comment]
Wow, that's an amazingly interesting point, @luisherranz. So, if I get it right—regardless of the whole Interactivity API context—in Gutenberg you can never really rely on white/blacklisting block children because editing a pattern/template part could be used as a work around that limitation? If that's the case, this seems to me to be a pretty big issue with GB itself: is that documented? Is it a known-issue/looking for fix, or a `wontfix` kind of situation?

Also, just slightly related question: maybe I am not up to date with the latest developments of the Patterns, but last time I used them, patterns would just insert a series of blocks and, once inserted, editing the pattern would not have any effect on the already inserted one (unlike the Template Parts). Has this changed?

**Comment by luisherranz** [comment]
> in Gutenberg you can never really rely on white/blacklisting block children because editing a pattern/template part could be used as a work around that limitation? 

That's right. In the Editor, you can get the list of blocks that are inside those Patterns/Template Parts when you are editing the template, but nothing prevents users from modifying those Patterns/Template parts outside of that template.

> If that's the case, this seems to me to be a pretty big issue with GB itself: is that documented?

I don't think so. It's not that it's right or wrong, it's just the way it works.

> Is it a known-issue/looking for fix, or a wontfix kind of situation?

As far as I know, nobody is looking to fix or improve that. I'm not even sure it's fixable at all.

**Comment by kmanijak** [comment]
> last time I used them, patterns would just insert a series of blocks and, once inserted, editing the pattern would not have any effect on the already inserted one (unlike the Template Parts). Has this changed?

_Patterns_ still work that way but there are also [Synced Patterns (aka Reusable Blocks)](https://wordpress.org/documentation/article/reusable-blocks/) and that can change dynamically.

---

I'm more and more keen to follow Luis' suggestion with three options:
> In a way, this aligns better with our initial ideas about how client-side navigation should work in WordPress, with three settings:
> `off`: it's always turned off
> `auto`: it's turned on if everything is compatible (that's how it currently works in the Query block when you enable it)
> `on`: it's always turned on, use it at your own risk

Although, I'd love it to be a seamless experience. One thing is allowing them to change that setting but another thing is explaining to users what "blocks incompatible with Interactivity API" mean and in what circumstances they should turn the navigation off 😞 Still need to think it through.

**Comment by luisherranz** [comment]
My two cents:

If you want it to be as seamless as possible, maybe you could use `auto` (no setting to turn it `off`) AND prevent the insertion of non-compatible blocks in the Editor.

Then, the only way for users to disable client-side navigation is to insert a pattern without non-compatible blocks inside the Product Collection block, save the template, edit the pattern in isolation (or from another template), and add a non-compatible block. That's going to be a rare case so it should not happen very often.

Then, if that happened and people wonder why the Product Collection block is not doing client-side navigation anymore they will likely load the template to take a look. Once they load the template, you can detect that their Products Collection block contains a non-compatible block and alert the user about the problem, inviting them to delete the block.

**Comment by kmanijak** [comment]
Thanks @luisherranz! I like the solution you suggested. I actually started to implement that part:

> prevent the insertion of non-compatible blocks in the Editor.

But then I had a discussion with @gigitux about how much that affects extensibility which should be one of the key selling points of the Product Collection. Hence, we think we should skip that step and allow to add anything in Product Collection inner blocks.

However, we should implement all the other steps you mentioned:
1. Disable client-side navigation if the incompatible block is detected, no matter how it was added
2. Notify users that the client-side navigation will be disabled if the incompatible block is detected in Editor

Those steps would be a temporary measure until Interactivity API is public in WordPress 6.5 and until we propagate a way to mark blocks compatible/incompatible with Interactivity API but that will take time until block creators start doing that.

What do you think @luisherranz , @sunyatasattva , @gigitux , @imanish003 . Do you agree with that approach?

**Comment by gigitux** [comment]
Thanks, @kmanijak, for the ping! I think that it makes sense!

**Comment by luisherranz** [comment]
Then, it's going to be very similar to what we did for the Query block. You can take a look [at the code we used in the Editor](https://github.com/WordPress/gutenberg/blob/c726d853ebb710424dac17c1821080050e8d742a/packages/block-library/src/query/edit/enhanced-pagination-modal.js) for the modal messages and such, and [this is the PR that implemented the current functionality](https://github.com/WordPress/gutenberg/pull/55714).

If you want us to do an overview of how everything works in the Query block, let us know.

**Comment by kmanijak** [comment]
> Then, it's going to be very similar to what we did for the Query block. 

Yes, with the difference that we don't want to expose the controls in the Inspector Controls.

Actually, I should've done this earlier but @manospsyx , can you confirm my assumption? Here's the TLDR for you:
- we cannot use client-side navigation powered by Interactivity API if the non-compatible blocks are used within Product Collection
- eventually what we want to do is:
    - Disable client-side navigation if the incompatible block is detected, no matter how it was added
    - Notify users that the client-side navigation will be disabled if the incompatible block is detected in Editor
- Query Loop block exposes such control which gets disabled if the incompatible block is detected in Editor:
<img width="276" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/95b32a4a-b9a3-4379-b7d1-843ffa296d4d">

**My assumption:** We don't want to expose such control as it should work seamlessly without merchants worrying about that. Is that correct?
**Question:** Do you have an idea how should we notify users about disabling client-side navigation (force reload on page change).

> If you want us to do an overview of how everything works in the Query block, let us know.

@luisherranz thanks! I browsed your PRs and I think it all makes sense, I'll contact you if I encounter some troubles!

**Comment by luisherranz** [comment]
I’d like to document another case so you can also take it into account: blocks whose inner blocks can be different on each render.

The best example of this type of block is the Post Content block, but any block that can conditionally render different sets of inner blocks is problematic.

The issue is the same one that we explained for the “No results” page: if the new page contains blocks that didn’t exist on the previous page, the CSS and JS of those blocks are missing.

Imagine this scenario:

We load the first page of a Query block and it contains two posts:

- Post-content of post 1: Heading and paragraph blocks
- Post-content of post 2: Heading and image blocks

On the page load, Gutenberg loads the CSS for the heading, paragraph and image blocks.

We do a client navigation to page 2, which also contains two posts:

- Post-content of post 3: Image and paragraph
- Post-content of post 4: Heading and list

The Interactivity API loads the HTML of all those blocks, including the new list block. The image, paragraph and heading blocks are fine, but the CSS of the list block is missing and therefore it is unstyled.

That’s the reason we treat the Post Content block as non-compatible in the Query block.

**Comment by manospsyx** [comment]
@kmanijak I missed this -- better late than never.

> We don't want to expose such control as it should work seamlessly without merchants worrying about that. Is that correct? 

I'd love to hear from @jarekmorawski about this. I think we covered some of this already? 

I generally prefer to make decisions for the user, rather than give them options to figure out: The option does not serve a real user need (who wants to force page refreshes?), but it sounds like:

- There are scenarios where the shopper/visitor experience might appear inconsistent for reasons that they (or the site admin) may not understand... and then site admins may want to enable this to make the visitor/shopper experience more predictable? 
- When an incompatible block is used, then site admins / editors can be notified in the Editor. In this case, I would maybe opt to keep the option enabled in a "greyed out" state, and use the description to let them know why. If we view this as a "warning" state, then we could use a signifier to indicate that.

**Comment by kmanijak** [comment]
> @kmanijak I missed this -- better late than never.

No worries!

> I'd love to hear from @jarekmorawski about this. I think we covered some of this already?

We discussed that and we got a way forward (apparently, I didn't post an update here).

We'll follow Query Loop approach with the control like this that gets disabled when 3PD block is inserted in Product Collection:

![image](https://github.com/woocommerce/woocommerce/assets/20098064/88c21009-73b0-449a-ba96-1caac35cb2e1)

The control will be in Advanced section so it doesn't bring attention. Also, we aim to add this orange dot indicating if the option got disabled:

![image](https://github.com/woocommerce/woocommerce/assets/20098064/484bfe0c-340d-4409-9f08-d7086b4e548e)

Ref: p1705933850042949/1705681434.462809-slack-C05UMTA7W0P

**Comment by kmanijak** [comment]
For reference, there's issue in which plugin seems not to be block-based but still renders some stuff based on DOM events (after load). With the client-side navigation it won't detect page transitions: https://github.com/woocommerce/woocommerce/issues/44525.

One option suggested by user is to expose custom JavaScript event broadcasting information about page transition.

EDIT: I misunderstood the content of the issue. The extension is block-based and that's exactly an example of what's been discussed within this issue.

**Comment by DAreRodz** [comment]
I probably don't have the full context here, but I wanted to chime in. 👋 

In WP 6.5, developers should specify their blocks' compatibility with the Interactivity API with the `supports.interactivity` attribute in `block.json` (see [Block API Reference / Supports#Interactivity](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#interactivity)).

With that in mind, this is what you have to do to enable client-side navigation:

1. For interactive blocks, use `"interactivity": true`.
2. For non-interactive blocks, use `"interactivity": { "clientNavigation": true }`.

Note that these are required for client navigation because the default values are `false`. All Core blocks have them, but third-party plugins must update their blocks to mark them as "compatible" and enable the feature.

However, if what you want is to turn off the feature, you can modify the `core/router` store configuration using the following code in PHP ([like the Query block does](https://github.com/WordPress/gutenberg/blob/535252ede8a5ad48d425c68da64f3a44d0bd973a/packages/block-library/src/query/index.php#L141)):

```php
wp_interactivity_config(
    'core/router',
    array( 'clientNavigationDisabled' => true )
);
```

When navigating _from_ or _to_ a page with `clientNavigationDisabled` as `true`, the router forces the page to reload.

Hope it helps!

**Comment by kmanijak** [comment]
@DAreRodz, thanks a lot for this comment!

> In WP 6.5, developers should specify their blocks' compatibility with the Interactivity API with the supports.interactivity attribute in block.json (see [Block API Reference / Supports#Interactivity](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#interactivity)).

That means we should mark the blocks in WooCommerce! Nice!

> However, if what you want is to turn off the feature, you can modify the core/router store configuration using the following code in PHP ([like the Query block does](https://github.com/WordPress/gutenberg/blob/535252ede8a5ad48d425c68da64f3a44d0bd973a/packages/block-library/src/query/index.php#L141)):
When navigating from or to a page with clientNavigationDisabled as true, the router forces the page to reload.

Can we use `@wordpress/interactivity-router`? Docs says:

> [Can I use it?](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-interactivity-router/#can-i-use-it)
> You can test it, but it’s still very experimental.

Or am I missing something and we can use `core/router` store without `@wordpress/interactivity-router` package?

**Comment by DAreRodz** [comment]
> [Can I use it?](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-interactivity-router/#can-i-use-it)
> You can test it, but it’s still very experimental.

Oh, my. You can. We still have to update the README file... 🙈 😅 

By the way, I think I haven't explained myself properly in the previous message. It's not that specifying `"interactivity": true` or `"interactivity: { "clientNavigation": true }` everything works out of the box. The logic that checks for those properties runs inside the Query block, and when that prop is missing for any inner block, that logic adds `clientNavigationDisabled: true` to the `core/router` store configuration.

You can take a look at the code [here](https://github.com/WordPress/gutenberg/blob/535252ede8a5ad48d425c68da64f3a44d0bd973a/packages/block-library/src/query/index.php#L91-L170).

As the logic is inside the block, you would have to implement something similar in the Product Collection block.

In summary:
1. Blocks declare compatibility using the `interactivity` support.
2. The Query block (or the Product Collection) checks in PHP if there are incompatible blocks.
3. If that's the case, add `clientNavigationDisabled` to the `core/router` config with `wp_interactivity_config()`.

**Comment by kmanijak** [comment]
> By the way, I think I haven't explained myself properly in the previous message. It's not that specifying "interactivity": true or "interactivity: { "clientNavigation": true } everything works out of the box. The logic that checks for those properties runs inside the Query block, and when that prop is missing for any inner block, that logic adds clientNavigationDisabled: true to the core/router store configuration.

> You can take a look at the code [here](https://github.com/WordPress/gutenberg/blob/535252ede8a5ad48d425c68da64f3a44d0bd973a/packages/block-library/src/query/index.php#L91-L170).

Ah yes, I've already started to "mimic" that in Product Collection block!

> That means we should mark the blocks in WooCommerce! Nice!

What I meant is WooCommerce blocks should set the example by declaring (or not) compatibility as a 3PD blocks.

**Comment by DAreRodz** [comment]
> What I meant is WooCommerce blocks should set the example by declaring (or not) compatibility as a 3PD blocks.

Yup. WooCommerce blocks should declare compatibility. 😄

**Comment by kmanijak** [comment]
@jarekmorawski , I won't be able to implement the orange dot notification next to "Advanced". I did give it a go and unfortunately:
- of course, there's no such feature in Gutenberg which we knew upfront,
- I tried to implement it "manually" and while I can cover most of the cases, it's getting really hacky and has a performance impact on the Editor. I know it's just an orange dot but unfortunately, I have to manipulate the DOM directly which conflicts with a React and it's a really bad idea to pursue that path.

Having said that, do you have any other suggestions? In Gutenberg, when the incompatible block is detected, they display the modal explaining the situation but it's quite invasive and we wanted to avoid that.

**Comment by asafm7** [comment]
Hello, @kmanijak .

As forcing a page reload might compromise the performance and user experience, I was wondering if the option of exposing custom JS events as described on https://github.com/woocommerce/woocommerce/issues/44525 was considered.

**Comment by asafm7** [comment]
Is there a way to bypass this?

![image](https://github.com/woocommerce/woocommerce/assets/14274769/7b7a0ade-d01f-4f57-98a3-e762e0255479)

**Comment by poof86** [comment]
> Is there a way to bypass this?

@asafm7 they are solving this in the next gutenberg release:  https://github.com/WordPress/gutenberg/issues/59752

**Comment by asafm7** [comment]
Thanks, @poof86.

Just making sure we are on the same page: in my case, there is a block that doesn't declare Interactivity support (YITH WooCommerce Wishlist). Even so, I've coded a JS workaround to make it work. In this case, will it be possible to bypass the Force Page Reload setting? (I'm not sure it is similar to the case described in the ticket you referenced).

Thanks again.

**Comment by poof86** [comment]
Oh I think they are fixing it only for blocks that declare interactivity support, so it's indeed a little different. I extended the Query Block with a custom button to forcefully enable/disable force reload. I then use this value set the "enhancedPagination" attribute to "true" using the "register_block_type_args" filter.
